firedrake/adjoint/ensemble_reduced_functional.py

```diff
@@ -53,15 +61,72 @@ class EnsembleReducedFunctional(ReducedFunctional):
     works, please refer to the `Firedrake manual
     <https://www.firedrakeproject.org/parallelism.html#id8>`_.
     """
-    def __init__(self, J, control, ensemble, scatter_control=True):
-        super(EnsembleReducedFunctional, self).__init__(J, control)
+    def __init__(self, J, control, ensemble, scatter_control=True,
+                 gather_functional=None):
         self.ensemble = ensemble
+        if isinstance(J, list):
+            assert isinstance(control, list), "Controls should also be a list."
+            assert len(control) % len(J) == 0, "Controls and J have mismatching lengths."
+            self.controls_per_J = len(control) // len(J)
+            self.functional = J
+            self.controls = Enlist(control)
+            self.Jhats = []
+            self.sizes = self.ensemble.ensemble_comm.allgather(len(J))
+            for i in range(len(J)):
+                i0 = self.controls_per_J*i
+                i1 = self.controls_per_J*(i+1)
+                self.Jhats.append(ReducedFunctional(J[i], control[i0:i1]))
+        else:
+            super(EnsembleReducedFunctional, self).__init__(J, control)
         self.scatter_control = scatter_control
+        self.gather_functional = gather_functional
+
+    def _allgather_J(self, J):
+        if isinstance(J, list):
+            rank = self.ensemble.ensemble_comm.rank
+            vals = []
+            # allgather a flattened list of all of the functional values
+            for i in range(len(self.sizes)):
+                for j in range(self.sizes[i]):
+                    Jtype = type(J[j]) if i == rank else None
+                    Jtype = self.ensemble.ensemble_comm.bcast(Jtype, root=i)
+                    if issubclass(Jtype, float):
+                        Jsend = J[j] if i == rank else None
+                        vals.append(self.ensemble.ensemble_comm.bcast(Jsend,
+                                                                      root=i))
+                    elif issubclass(Jtype, firedrake.Function):
+                        Jsend = J[j].copy(deepcopy=True) if i == rank else None
+                        vals.append(self.ensemble.bcast(Jsend, root=i))
+                    else:
+                        raise NotImplementedError("This type of functional is not supported: " + str(Jtype))
+        elif isinstance(J, float):
```

Comment:
if you always enlist J, you lose the last two cases here.

Reply:
Where am I enlisting J?

Reply:
you;re not. that's the issue. All over pyadjoint they use this enlist trick to deal with exactly this case: situations where either one item or a list would be OK.